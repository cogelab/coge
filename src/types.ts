export interface Logger {
  ok: (msg: string) => void
  notice: (msg: string) => void
  warn: (msg: string) => void
  err: (msg: string) => void
  log: (msg: string) => void
  colorful: (msg: string) => void
}

type CliCmdValidatorFn = (str: string) => any;
type CliCmdValidatorArg = string[] | string | RegExp | CliCmdValidatorFn | Number;

type CliCmdActionCallback = (opts: CmdOptions) => Promise<any>;

type CliCmdOptionType = 'int' | 'float' | 'bool' | 'list' | 'repeatable';

export interface CliCmdArgument {
  flags: string;
  description: string;
  type?: CliCmdOptionType;
  validator?: CliCmdValidatorArg;
  defaultValue?: any;
}

export interface CliCmdOption {
  flags: string;
  description: string;
  type?: CliCmdOptionType;
  validator?: CliCmdValidatorArg;
  defaultValue?: any;
  required?: boolean;
}

export interface CliCmdDefinition {
  name: string;
  description: string;
  default?: boolean;
  alias?: string;
  help?: string;
  arguments?: CliCmdArgument[];
  options?: CliCmdOption[];
  action: CliCmdActionCallback;
}

export interface CliOptions extends Partial<RunnerConfig> {
  logger?: Logger;
}

export interface Loader {
  exists: (arg0: string) => Promise<boolean>
  load: (arg0: string) => Promise<Record<string, any>>
  none: (arg0: string) => Record<string, any>
}

export interface Prompter<Q, T> {
  prompt: (arg0: Q) => Promise<T>
}

export interface RenderedAction {
  file?: string
  attributes: any
  body: string
}

export interface RunnerConfig {
  cwd: string
  templates: string
  exec: (sh: string, body: string) => void
  debug: boolean
  helpers: any
  createPrompter: <Q, T>() => Prompter<Q, T>
}

export type Params = {
  templates: string;
  generator: string;
  action: string;
  folder?: string;
  pattern?: string;
  dry?: boolean;
  // name?: string;
}

export type CmdOptions = {
  conf: RunnerConfig;
  opts: { [k: string]: any };
  logger: Logger;
}

export type OpResult = any
export type Op = (action: RenderedAction, opts: any, conf: RunnerConfig, logger: Logger) => Promise<OpResult>

