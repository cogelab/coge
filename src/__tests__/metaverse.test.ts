import {createLogger} from "./support";
import {CliOptions} from "../types";

const path = require('path')
const dirCompare = require('dir-compare')

const opts = {compareContent: true}
const fs = require('fs-extra')
const enquirer = require('enquirer')
const {cli} = require('../cli')

const logger = createLogger();
const failPrompt = () => {
  throw new Error('set up prompt in testing')
}

jest.mock('enquirer', () => ({
  prompt: null,
}))

const SKIP_ON_WINDOWS = process.platform === 'win32' ? ['shell'] : [];

const dir = m => path.join(__dirname, 'metaverse', m);
const run = async (argv: any[], options: CliOptions) => cli(['node', 'coge', ...argv], options);
const metaverse = (folder, cmds, promptResponse) => it(folder, async () => {
  const metaDir = dir(folder)
  console.log('metaverse test in:', metaDir)
  const config = {
    templates: '_templates',
    cwd: metaDir,
    exec: (action, body) => {
      const opts = body && body.length > 0 ? {input: body} : {}
      return require('execa').shell(action, opts)
    },
    logger,
    createPrompter: () => require('enquirer'),
  }
  // await fs.remove(path.join(metaDir, 'given'))
  console.log('before', fs.readdirSync(metaDir))
  for (let cmd of cmds) {
    console.log('testing', cmd)
    if (
      process.platform === 'win32' &&
      SKIP_ON_WINDOWS.find(c => cmd[0] === c)
    ) {
      console.log(`skipping ${cmd} (windows!)`)
      await fs.remove(path.join(metaDir, 'expected', cmd[0]))
      continue
    }

    enquirer.prompt = failPrompt
    if (promptResponse) {
      const last = cmd[cmd.length - 1]
      if (typeof last === 'object') {
        cmd = cmd.slice(0, cmd.length - 1)
        enquirer.prompt = () => Promise.resolve({...promptResponse, ...last})
      } else {
        enquirer.prompt = () => Promise.resolve(promptResponse)
      }
    }
    const res = await run(cmd, config)
    res.actions.forEach(a => {
      a.timing = -1
      a.subject = a.subject.replace(/.*coge\/src/, '')
    })
    expect(res).toMatchSnapshot(cmd.join(' '))
  }
  const givenDir = path.join(metaDir, 'given')
  const expectedDir = path.join(metaDir, 'expected')
  console.log('after', {
    [givenDir]: fs.readdirSync(givenDir),
    [expectedDir]: fs.readdirSync(expectedDir),
  })
  const res = dirCompare.compareSync(givenDir, expectedDir, opts)
  res.diffSet = res.diffSet.filter(d => d.state !== 'equal')
  if (!res.same) {
    console.log(res)
  }
  expect(res.same).toEqual(true)
})

describe.only('metaverse', () => {
  beforeAll(() => {
    fs.removeSync(dir('coge-extension/given'));
    fs.removeSync(dir('coge-templates/given'));
  })
  beforeEach(() => {
    enquirer.prompt = failPrompt
  })
  metaverse('coge-extension', [['coge-js', 'new']], {overwrite: true})
  metaverse(
    'coge-templates',
    [
      ['init', 'self'],
      ['overwrite-yes', 'base'],
      ['overwrite-yes', 'over'],
      ['overwrite-no', 'base'],
      ['overwrite-no', 'over', {overwrite: false}],
      ['mailer', 'new'],
      ['worker', 'new', '--name', 'foo'],
      ['shell', 'new', '--name', 'foo'],
      ['inflection', 'new', '--name', 'person'],
      ['conditional-rendering', 'new', '-a notGiven'],
      ['add-unless-exists', 'new', '-a message=foo'],
      [
        'cli-prefill-prompt-vars',
        'new',
        '-a message-from-cli=hello-from-cli',
      ],
      ['cli-prefill-prompt-vars', 'name-is-special', 'foobar'],
      [
        'cli-prefill-prompt-vars',
        'falsy-values-are-ok',
        'foobar',
        '-a include_something=false',
      ],
      ['recursive-prompt', 'new'],
      ['positional-name', 'new', 'acmecorp'],
      ['existing-params', 'new', '-a email=premade-email@foobar.com'],
      [
        'existing-params',
        'new-params-alias',
        '-a email=premade-email@foobar.com',
      ],
      [
        'index-js-existing-params',
        'new',
        '-a email=premade-email@foobar.com',
      ],
      [
        'index-js-existing-params',
        'new-params-alias',
        '-a email=premade-email@foobar.com',
      ],
    ], // this is all of the responses enquirer gives out from _all_ tests, ever.
    // it's best to just keep it that way to be simple, and each prompt-dealing test
    // has its own set of uniquely named variables.
    {
      // generic for all tests
      name: 'message',
      message: 'foo',
      overwrite: true,

      // recursive-prompt
      email: 'some-email@foobar.com',
      emailConfirmation: 'confirmed-some-email@foobar.com',
    },
  )
})
