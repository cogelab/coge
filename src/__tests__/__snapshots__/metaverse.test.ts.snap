// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`metaverse coge-extension: coge-js:new 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "generated",
      "subject": "given/coge-js/new.md",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse coge-generators: add-unless-exists:new -D message=foo 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "generated",
      "subject": "given/add-unless-exists/foo/always.txt",
      "timing": -1,
      "type": "add",
    },
    Object {
      "status": "generated",
      "subject": "given/add-unless-exists/foo/always.txt",
      "timing": -1,
      "type": "add",
    },
    Object {
      "status": "skipped",
      "subject": "given/add-unless-exists/foo/always.txt",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse coge-generators: cli-prefill-prompt-vars:falsy-values-are-ok foobar -D include_something=false 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "generated",
      "subject": "given/cli-prefill-prompt-vars/falsy.md",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse coge-generators: cli-prefill-prompt-vars:name-is-special foobar 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "generated",
      "subject": "given/cli-prefill-prompt-vars/name-is-special.md",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse coge-generators: cli-prefill-prompt-vars:new -D message-from-cli=hello-from-cli 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "generated",
      "subject": "given/cli-prefill-prompt-vars/new.md",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse coge-generators: conditional-rendering:new -D notGiven 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "generated",
      "subject": "given/conditional/always.txt",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse coge-generators: existing-params:new -D email=premade-email@foobar.com 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "generated",
      "subject": "given/existing-params/new.md",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse coge-generators: existing-params:new-params-alias -D email=premade-email@foobar.com 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "generated",
      "subject": "given/existing-params/new-params-alias.md",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse coge-generators: index-js-existing-params:new -D email=premade-email@foobar.com 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "generated",
      "subject": "given/index-js-existing-params/new.md",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse coge-generators: index-js-existing-params:new-params-alias -D email=premade-email@foobar.com 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "generated",
      "subject": "given/index-js-existing-params/new-params-alias.md",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse coge-generators: inflection:new --name person 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "generated",
      "subject": "given/my_app/people.md",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse coge-generators: init:self 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "generated",
      "subject": "given/templates/sample/new/hello.ejs.t",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse coge-generators: mailer:new 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "generated",
      "subject": "given/app/mailers/message.js",
      "timing": -1,
      "type": "add",
    },
    Object {
      "status": "generated",
      "subject": "given/app/mailers/foo/html.ejs",
      "timing": -1,
      "type": "add",
    },
    Object {
      "status": "generated",
      "subject": "given/app/mailers/foo/subject.ejs",
      "timing": -1,
      "type": "add",
    },
    Object {
      "status": "generated",
      "subject": "given/app/mailers/foo/text.ejs",
      "timing": -1,
      "type": "add",
    },
    Object {
      "status": "inject",
      "subject": "given/app/mailers/foo/html.ejs",
      "timing": -1,
      "type": "inject",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse coge-generators: overwrite-no:base 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "generated",
      "subject": "given/overwrite-no/new.md",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse coge-generators: overwrite-no:over 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "skipped",
      "subject": "given/overwrite-no/new.md",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse coge-generators: overwrite-yes:base 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "generated",
      "subject": "given/overwrite-yes/new.md",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse coge-generators: overwrite-yes:over 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "generated",
      "subject": "given/overwrite-yes/new.md",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse coge-generators: positional-name:new acmecorp 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "generated",
      "subject": "given/positional-name/acmecorp/always.txt",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse coge-generators: recursive-prompt:new 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "generated",
      "subject": "given/recursive-prompt/new.md",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse coge-generators: shell:new --name foo 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "executed",
      "subject": "/__tests__/metaverse/coge-generators/given/shell/hello.piped",
      "timing": -1,
      "type": "shell",
    },
    Object {
      "status": "executed",
      "subject": "/__tests__/metaverse/coge-generators/given/shell/foo.hello",
      "timing": -1,
      "type": "shell",
    },
    Object {
      "status": "generated",
      "subject": "given/shell/hmm.txt",
      "timing": -1,
      "type": "add",
    },
    Object {
      "status": "executed",
      "subject": "/__tests__/metaverse/coge-generators/given/shell/side-effect.hello",
      "timing": -1,
      "type": "shell",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse coge-generators: worker:new --name foo 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "generated",
      "subject": "given/app/workers/foo.js",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;
